### ISSUE TITLE
observables not working in a release build

### ISSUE THREAD

**Issue by alekes** [mobxjs/mobx-react#444]
I struggle since several days with a problem and can't get find a solution :-(
The following code snippets are part of a bigger React Native app which uses observables all over (and they're working fine), but this part doesn't work when I build a release.
That's right - Everything is working fine in a debug build on Android and iOS but fails when I build a release version (no matter if the release is deployed to a device or simulator/emulator).

App.js
```
import React, { Component } from 'react';
import { Provider } from 'mobx-react';
import Store from './Store';
import Home from './Home';

const store = new Store();

export default class App extends Component {
  
  render() {
    return (
      <Provider store={store} >
        <Home />
      </Provider>
    );
  }
}

```

Home.js
```
import React, { Component } from 'react';
import { Text, View, Button } from 'react-native';
import { observer, inject } from "mobx-react/native";

@inject('store')
@observer
export default class Home extends Component {

  render() {
    return (
        <View>
          <Text>
            {this.props.store.text}
          </Text>
          <Button title="Add text" onPress={() => this.props.store.addText('text')}/>
        </View>
    );
  }
}
```

Store.js
```
import { observable, action } from "mobx";

export default class Store {
    @observable text = [];
    @action addText(value) {
        this.text.push(value)
    }
}
```

Pressing the "Add text" button will add "text" to the text array. Like I already wrote, this works as expected in DEV mode (no errors). Compiling a release build (signed apk or ipa) and it stops working (again no errors, the text is just not added).
The Home component doesn't seem to be updated. I can force that by adding a `setState` call which is triggered at the same time as `this.props.store.addText('text')`. After that it also works in a release build, so it doesn't seem to be a minify problem of the build itself.

Any help is much appreciated.

**Comment by mweststrate** [comment]
Since this works fine during dev it is most probably a build setup issue. Can you provide a minimal reproduction project? Make sure that `mobx` is only once in your node modules (`tree -d node_modules | grep mobx`)

**Comment by alekes** [comment]
Thank you for your fast response!

Tree returns this (looks good):
```
tree -d node_modules | grep mobx
├── mobx
├── mobx-react

```

I created the following repo [https://github.com/alekes/mobxtest](https://github.com/alekes/mobxtest) with the example code.

Also included a dummy keystore to build the Android version right away.

Building a debug version everything works fine:
`react-native run-android`

Building a release version the described behavior will take effect:
`react-native run-android --variant=release`

Hint: I did a test checkout of that repo and the Android build wasn't working anymore. A simple `react-native upgrade` did fix it.

**Comment by mweststrate** [comment]
@alekes I noticed you are using an alpha version of react, does downgrading to 16.2 make a difference?

**Comment by alekes** [comment]
Downgrading react to the latest stable release didn't solve the problem.

I created an Expo project and get the same results.
The code runs fine on my computer but stops working after uploading to Expo:
[https://expo.io/@om-ak/mobxtestexpo](https://expo.io/@om-ak/mobxtestexpo)

Could it be a code minify/uglify thing?

Expo packages.json

 ```
"expo": "^25.0.0",
    "mobx": "^4.1.0",
    "mobx-react": "^5.0.0",
    "react": "16.2.0",
    "react-native": "https://github.com/expo/react-native/archive/sdk-25.0.0.tar.gz"

```

**Comment by alekes** [comment]
Today I tried to compile a release version of [the test repo I did commit]( https://github.com/alekes/mobxtest) on a different computer and ended up with the same result. Debug version works, compiled release version fails.

@mweststrate When I change my Home.js file and include a RN state call, also the this.props.store.text observable is updated in the release version.

Home.js (including setState):

```
import React, { Component } from 'react';
import { Text, View, Button } from 'react-native';
import { observer, inject } from "mobx-react/native";

@inject('store')
@observer
export default class Home extends Component {

  constructor() {
    super()
    this.state = {
      isUpdated: false,
    }
  }

  onButtonPress() {
    this.props.store.addText('text');
    this.setState({isUpdated: true});
  }

  render() {
    return (
        <View>
          <Text>
            {this.props.store.text}
          </Text>
          <Button title="Add text" onPress={() => this.onButtonPress()}/>
        </View>
    );
  }
}
```
This makes me believe that something might be wrong with my MobX code and that's the reason why the Home.js component doesn't update.

Is this maybe a problem of the array.push call I use inside Store.js?

```
 @action addText(value) {
        this.text.push(value)
    }
```

**Comment by mweststrate** [comment]
@alekes just checked, but you are rendering an array into a react component: `{this.props.store.text}`, that is something React can't really do anything with, probably you want do something like `{this.props.store.text.join("")}`. Not sure whether it makes a difference between the builds, but I suspect you should have warnings from react about that in any case

**Comment by AntonPuko** [comment]
I probably have similar issue. not sure if its related to mobx or mobx-react, but: 
I have a store that looks like that:

```
type Field = {
 error?: Error,
};

class FormStore {
  @obsrervable _fields: ObservableMap<string, FIeld> = new Map();

  @computed
  get invalid(): boolean {
    return Array.from(this._fields.values()).some(
      (field: Field): boolean => !!field.error,
    );
  }
}
```

Everything works okay when `debug js` is on, but problems start when debugging mode disabled:
1) If there are 2 or fewer fields inserted in the map, everything works okay.
2) with 3+ fields in the map: computed computation fails with the error: `TypeError: undefined is not an object(evaluation 'field.error')`
3) if I transform map using `.toJS()` like this: `Array.from(this._fields.toJS().values())` it starts working again in all cases. 

mobx 4.1.0
mobx-react 5.0.0
react-native 0.53.0
react: 16.0.0

Didn't create separate issue yet because It looks very subtle and I had no idea how to describe it properly.

**Comment by binyar** [comment]
I may have similar issue with you

> TypeError: (0 , _mobx.configure) is not a function

you can't use the latest version of `mobx-react`(5.0.0),when I use `4.1.0`，he problem was solved.

    "mobx": "*",
    "mobx-react": "^4.1.0",
    "react": "^16.2.0",
    "react-dom": "^16.2.0",

**Comment by mweststrate** [comment]
@AntonPuko that probably means that you are rendering observables in a component that doesn't have `observer`, or that you are not rendering a complex observable, but merely dumping it into react (e.g. `{this.myObservableArray}` instead of mapping over it and actually rendering the values.

**Comment by mweststrate** [comment]
@binyar yes, you have to combine the correct major versions. You should get peer dependency warnings if you did it wrong. MobX 4 goes with mobx-react 5, Mobx 3 with mobx-react 4. Also, don't use use `"mobx": "*"`! Or I will blow up your project every few months. Specify an actual range, like `^3` ,or `^4`

**Comment by mweststrate** [comment]
Closing this issue now, as it is discussing 3 quite different problems, and not making it more confusing. For any follow up questions, open a separate issue.
